Fix bulk index detailed stats and log failures

b-deam opened the pull request:
This commit checks and decodes bulk request bodies to UTF-8 
encoded strings, allowing users to now set the `detailed-results` operation 
property. As an added bonus we now also log individual bulk failures, as 
users previously had to either, enable `DEBUG` logging and navigate through
all Elasticsearch requests, or use a remote metrics store and check the 
stats document manually via Discover.

## Track details:

I built a test track with my own lab data:

- [Fail-bulk Track (including docs)](https://github.com/elastic/rally/files/7701477/fail-bulk.zip)

## Expected results
```
$ esrally race --distribution-version=7.16.0 --track-path=~/perf/rally/tracks/fail-bulk --kill-running-processes

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Race id is [6c86e9a0-5f86-4477-a964-13ddc33b7a22]
[INFO] Preparing for race ...
[INFO] Racing on track [fail-bulk], challenge [2-bulk-failures] and car ['defaults'] with version [7.16.0].

Running bulk                                                                   [100% done]

------------------------------------------------------
    _______             __   _____
   / ____(_)___  ____ _/ /  / ___/_________  ________
  / /_  / / __ \/ __ `/ /   \__ \/ ___/ __ \/ ___/ _ \
 / __/ / / / / / /_/ / /   ___/ / /__/ /_/ / /  /  __/
/_/   /_/_/ /_/\__,_/_/   /____/\___/\____/_/   \___/
------------------------------------------------------
            
|                                                         Metric |   Task |      Value |   Unit |
|---------------------------------------------------------------:|-------:|-----------:|-------:|
|                     Cumulative indexing time of primary shards |        |   0.137117 |    min |
|             Min cumulative indexing time across primary shards |        |    0.01645 |    min |
|          Median cumulative indexing time across primary shards |        |  0.0685583 |    min |
|             Max cumulative indexing time across primary shards |        |   0.120667 |    min |
|            Cumulative indexing throttle time of primary shards |        |          0 |    min |
|    Min cumulative indexing throttle time across primary shards |        |          0 |    min |
| Median cumulative indexing throttle time across primary shards |        |          0 |    min |
|    Max cumulative indexing throttle time across primary shards |        |          0 |    min |
|                        Cumulative merge time of primary shards |        |          0 |    min |
|                       Cumulative merge count of primary shards |        |          0 |        |
|                Min cumulative merge time across primary shards |        |          0 |    min |
|             Median cumulative merge time across primary shards |        |          0 |    min |
|                Max cumulative merge time across primary shards |        |          0 |    min |
|               Cumulative merge throttle time of primary shards |        |          0 |    min |
|       Min cumulative merge throttle time across primary shards |        |          0 |    min |
|    Median cumulative merge throttle time across primary shards |        |          0 |    min |
|       Max cumulative merge throttle time across primary shards |        |          0 |    min |
|                      Cumulative refresh time of primary shards |        |    0.13505 |    min |
|                     Cumulative refresh count of primary shards |        |         15 |        |
|              Min cumulative refresh time across primary shards |        |  0.0454667 |    min |
|           Median cumulative refresh time across primary shards |        |   0.067525 |    min |
|              Max cumulative refresh time across primary shards |        |  0.0895833 |    min |
|                        Cumulative flush time of primary shards |        |    0.00435 |    min |
|                       Cumulative flush count of primary shards |        |          1 |        |
|                Min cumulative flush time across primary shards |        |          0 |    min |
|             Median cumulative flush time across primary shards |        |   0.002175 |    min |
|                Max cumulative flush time across primary shards |        |    0.00435 |    min |
|                                        Total Young Gen GC time |        |      0.016 |      s |
|                                       Total Young Gen GC count |        |          3 |        |
|                                          Total Old Gen GC time |        |          0 |      s |
|                                         Total Old Gen GC count |        |          0 |        |
|                                                     Store size |        |  0.0381134 |     GB |
|                                                  Translog size |        |  0.0356154 |     GB |
|                                         Heap used for segments |        |  0.0659294 |     MB |
|                                       Heap used for doc values |        |  0.0160408 |     MB |
|                                            Heap used for terms |        |  0.0402832 |     MB |
|                                            Heap used for norms |        | 0.00524902 |     MB |
|                                           Heap used for points |        |          0 |     MB |
|                                    Heap used for stored fields |        | 0.00435638 |     MB |
|                                                  Segment count |        |          9 |        |
|                                                 Min Throughput |   bulk |      26.07 | docs/s |
|                                                Mean Throughput |   bulk |      49.95 | docs/s |
|                                              Median Throughput |   bulk |      37.53 | docs/s |
|                                                 Max Throughput |   bulk |      98.68 | docs/s |
|                                        50th percentile latency |   bulk |    188.454 |     ms |
|                                        90th percentile latency |   bulk |    352.396 |     ms |
|                                       100th percentile latency |   bulk |    3802.71 |     ms |
|                                   50th percentile service time |   bulk |    188.454 |     ms |
|                                   90th percentile service time |   bulk |    352.396 |     ms |
|                                  100th percentile service time |   bulk |    3802.71 |     ms |
|                                                     error rate |   bulk |      27.27 |      % |

[WARNING] Error rate is 27.27 for operation 'bulk'. Please check the logs.

--------------------------------
[INFO] SUCCESS (took 48 seconds)
--------------------------------

# ~/.rally/logs/rally.log
2021-12-13 06:16:43,792 ActorAddr-(T|:55288)/PID:97165 esrally.driver.runner WARNING Bulk request failed: [HTTP status: 400, message: failed to parse field [apparent_t] of type [float] in document with id '0Qxvsn0BNTFxaU2uqxEl'. Preview of field's value: 'ThisWontWork']
2021-12-13 06:16:44,147 ActorAddr-(T|:55288)/PID:97165 esrally.driver.runner WARNING Bulk request failed: [HTTP status: 400, message: failed to parse field [wind_spd_kmh] of type [long] in document with id 'YAxvsn0BNTFxaU2usBIY'. Preview of field's value: 'ThisWontWork']
```

Closes https://github.com/elastic/rally/issues/1393

DJRickyB reviewed (commented):
Do we still care about the `str` and `list` branch of this code? 

If we always anticipate `bytes` we should be able to remove some dead code, and also calculate the doc size from the request if `with_action_metadata` is `False`, rather than eating the "re-encode and count the bytes" overhead on a per-line basis.

pquentin commented:
Also, is it easy to test this explicitly?

b-deam commented:
> Do we still care about the `str` and `list` branch of this code?
> 
> If we always anticipate `bytes` we should be able to remove some dead code, and also calculate the doc size from the request if `with_action_metadata` is `False`, rather than eating the "re-encode and count the bytes" overhead on a per-line basis.

Good pickup. I refactored to handle this, as far as I can see we only ever expect the params body to be of type `bytes`:
https://github.com/elastic/rally/blob/7460e31f79c3a9af80f554bef6cbe238b11b98fb/esrally/track/params.py#L1235-L1255


> Also, is it easy to test this explicitly?

Yep, I've added a unit test.

b-deam commented:
@elasticsearchmachine test this please

pquentin commented:
@elasticsearchmachine test this please

pquentin commented:
I addressed the first failure in #1481, but now this fails with the infamous ` assert 16384 == 0`.

@elasticsearchmachine please test this again

pquentin commented:
@elasticsearchmachine test this please

b-deam commented:
Thanks for fixing the tests @pquentin!

pquentin reviewed (approved):
Thanks! I like that we don't have to decode the bytes! I'm going to merge this to make sure this get tested in the next night-rally round.

dliappis commented on esrally/driver/runner.py:
```diff
@@ -520,15 +520,11 @@ def detailed_stats(self, params, response):
 
-        for line_number, data in enumerate(bulk_lines):
-            line_size = len(data.encode("utf-8"))
+        for line_number, data in enumerate(params["body"].split(b"\n")):
```
This seems like a lighter thing to do, but since we are in the hot code path, did you check the performance impact between the two approaches?

pquentin replied:
Is a microbenchmark enough?

```
∮ python -m timeit -s 's = "cafécafé\n".encode() * 1000' 's.split(b"\n")'
10000 loops, best of 5: 22.7 usec per loop
∮ python -m timeit -s 's = "cafécafé\n".encode() * 1000' 's.decode().split("\n")'
5000 loops, best of 5: 44.7 usec per loop
```

Note that we can´t really compare the previous approach because it would just raise an exception, as `params["body"]` became bytes at some point and `detailed-results` was thus broken until now.

dliappis replied:
Yes ofc a microbenchmark for such cases is always very insightful. I haven't spent the time to go through all the places where bulk reads from a file, I was curious to see that we've removed the `data.encode('utf-8')` here, but added an `encode` in https://github.com/elastic/rally/pull/1400/files#diff-026a25af4075b17db2736adf14aa844c0aaeb2bc995b4b4ae94c432b0380b51fR350, hence wondering if we've measured the impact of the combined changes.

pquentin replied:
This encode is in tests, though

dliappis replied:
Ah yes ofc thank you! Sorry for the noise then and thanks for the microbenchmark.
